### ISSUE TITLE
Compatibility of DQN and parallel environments

### ISSUE THREAD

**Issue by maxwab** [MillionIntegrals/vel#45]
Hi,

I have been trying to train a DQN on breakout using the example config proposed, but by using several environments instead of just 1 to speed up training (which I understood is possible since version 0.3, according to comments in issue #28 ).
I modified the `breakout_ddqn.yaml` configuration file by using the `vel.rl.vecenv.shared_mem` vec_env and set the parameter `parallel_envs`  of reinforcer to 4, 8 and 12 (file attached).

[breakout_ddqn_parallel.yaml.txt](https://github.com/MillionIntegrals/vel/files/3107903/breakout_ddqn_parallel.yaml.txt)

Once launched, the code indeed seems to create several environments (according to the processes created in htop).
Moreover, the number of frames at each epoch grows linearly with the number of parallel envs, as expected.
However, the DQN does not train faster: for instance, with the 12 parallel envs, the DQN has an average episode reward of ~1-2 after
 ~9M frames.

This makes me think that the number of frames displayed does not correspond to the number of frames actually seen.

I might have been doing something wrong, in that case I would be very grateful for any advice!

Apart from that, thanks a lot for making this very nice repository public, I look forward to see its further developments.

Thanks in advance for your help!

Maxime

**Comment by MillionIntegrals** [comment]
I haven't really check this configuration, but as far as the idea I had it should work. I'll check it in a free moment, what goes on there exactly.

In the meantime it may be good to tweak some other settings that implicitly also affect the rate of learning. Since with more envs the batch size is significantly bigger, you may try to increase the learning rate as well as decrease `target_update_frequency`. 

In the meantime, I'll double check for any bugs.

**Comment by MillionIntegrals** [comment]
You can try this config, to see if it works:

```yaml
name: 'breakout_ddqn'


env:
  name: vel.rl.env.classic_atari
  game: 'BreakoutNoFrameskip-v4'


vec_env:
  name: vel.rl.vecenv.shared_mem
  frame_history: 4  # How many stacked frames go into a single observation


model:
  name: vel.rl.models.q_model

  input_block:
    name: vel.modules.input.image_to_tensor

  backbone:
    name: vel.rl.models.backbone.nature_cnn
    input_width: 84
    input_height: 84
    input_channels: 4  # The same as frame_history


reinforcer:
  name: vel.rl.reinforcers.buffered_off_policy_iteration_reinforcer

  env_roller:
    name: vel.rl.env_roller.transition_replay_env_roller

    replay_buffer:
      name: vel.rl.buffers.circular_replay_buffer

      buffer_initial_size: 1_000 # How many samples we need in the buffer before we start using replay buffer
      buffer_capacity: 25_000

      # Because env has a framestack already built-in, save memory by encoding only last frames in the replay buffer
      frame_stack_compensation: true
      frame_history: 4  # How many stacked frames go into a single observation

    action_noise:
      name: vel.rl.modules.noise.eps_greedy

      epsilon:
        name: vel.schedules.linear_and_constant
        end_of_interpolation: 0.1
        initial_value: 1.0
        final_value: 0.1

  algo:
    name: vel.rl.algo.dqn

    double_dqn: true
    target_update_frequency: 1_000  # After how many batches to update the target network
    max_grad_norm: 0.5

    discount_factor: 0.99

  rollout_steps: 4 # How many environment steps (per env) to perform per batch of training
  training_steps: 32 # How many environment steps (per env) to perform per training round
  parallel_envs: 12 # Roll out only one env in parallel, just like in DeepMind paper


optimizer:
  name: vel.optimizers.rmsprop
  lr: 2.5e-3
  alpha: 0.95
  momentum: 0.95
  epsilon: 1.0e-1


commands:
  train:
    name: vel.rl.commands.rl_train_command
    total_frames: 1.1e7  # 11M
    batches_per_epoch: 250
```

I've scaled some other settings to match the increase in batch size. 

I've run it locally for a few epochs to see if there is any speedup vs the non-parallel version and the speedup was there, although it was more of 2x rather than 10x which means there are also other bottlenecks, but it's hard for me to profile it at the moment. 

I was not able to find any obvious problems with the code, but I'm interested if this modified config will work. It's good to try. I've implemented the possibility to try DQN with parallel envs as it sounded as a good idea to me and it unified various algorithms, but I've never really tried it so far.


Another question is how to speed up DQN even further on a single node, I'll take a look at it when I have a free moment, but free moments are scarce...

**Comment by MillionIntegrals** [comment]
I got it to average reward of 117 after 11M frames which is kind of ok, but not great. Probably more hyperparam tweaks are needed to get something better.

**Comment by maxwab** [comment]
Thanks for the configuration file! When I tested I also got an average reward of ~110 after 11M frames.
I'll try to play a bit with the parameters. If a combination proves to be successful, I'll keep you updated.

**Comment by MillionIntegrals** [comment]
I'll close this issue now.
